rp2040-hal/src/critical_section_impl.rs

```diff
@@ -0,0 +1,75 @@
+use core::sync::atomic::{AtomicU8, Ordering};
+
+struct RpSpinlockCs;
+critical_section::custom_impl!(RpSpinlockCs);
+
+/// Marker value to indicate no-one has the lock.
+///
+/// Initialising `LOCK_OWNER` to 0 means cheaper static initialisation so it's the best choice
+const LOCK_UNOWNED: u8 = 0;
+
+/// Indicates which core owns the lock so that we can call critical_section recursively.
+///
+/// 0 = no one has the lock, 1 = core0 has the lock, 2 = core1 has the lock
+static mut LOCK_OWNER: AtomicU8 = AtomicU8::new(LOCK_UNOWNED);
+
+/// Marker value to indicate that we already owned the lock when we started the `critical_section`.
+///
+/// Since we can't take the spinlock when we already have it, we need some other way to keep track of `critical_section` ownership.
+/// `critical_section` provides a token for communicating between `acquire` and `release` so we use that.
+/// If we're the outermost call to `critical_section` we use the values 0 and 1 to indicate we should release the spinlock and set the interrupts back to disabled and enabled, respectively.
+/// The value 2 indicates that we aren't the outermost call, and should not release the spinlock or re-enable interrupts in `release`
+const LOCK_ALREADY_OWNED: u8 = 2;
+
+unsafe impl critical_section::Impl for RpSpinlockCs {
+    unsafe fn acquire() -> u8 {
+        // Store the initial interrupt state and current core id in stack variables
+        let interrupts_active = cortex_m::register::primask::read().is_active();
+        // We reserved 0 as our `LOCK_UNOWNED` value, so add 1 to core_id so we get 1 for core0, 2 for core1.
+        let core = (*pac::SIO::ptr()).cpuid.read().bits() as u8 + 1_u8;
+        // Do we already own the spinlock?
+        if LOCK_OWNER.load(Ordering::Acquire) == core {
+            // We already own the lock, so we must have called acquire within a critical_section.
+            // Return the magic inner-loop value so that we know not to re-enable interrupts in release()
+            LOCK_ALREADY_OWNED
+        } else {
+            // Spin until we get the lock
+            loop {
+                // Need to disable interrupts to ensure that we will not deadlock
+                // if an interrupt enters critical_section::Impl after we acquire the lock
+                cortex_m::interrupt::disable();
+                // Read the spinlock reserved for `critical_section`
+                if (*pac::SIO::ptr()).spinlock31.read().bits() != 0 {
+                    // We just acquired the lock.
```

Comment:
I think a `core::sync::atomic::compiler_fence(Ordering::SeqCst)` needs to be added here and just before the spinlock is released, below.
Otherwise, the compiler is allowed to re-order instructions in an invalid way.
Then, the ordering of the accesses to LOCK_OWNER could be changed to `Ordering::Relaxed`, as proper synchronization between the cores is guaranteed by the spinlock. (There can still be concurrent reads and writes due to the read at the beginning of the function, which is not covered by the spinlock. So we need the atomic access. But in that case, the actual value read doesn't matter, so `Ordering::Relaxed` will be sufficient.)